### PR TITLE
Feature order process

### PR DESCRIPTION
**Pull request by MinaKhamesi**
-  enabled the application to allow different users see other chef's profiles 
- in case the user is seeing other chef's profile-aka they are not the owner of the profile- they have the ability to select items and add them to the cart
- a cart field added to user model and editCart, deleteItemFromCart, deleteCart, getCartInfo controllers added to backend with their corresponding routes. To make the data persistent.
- they will have a button to contact the chef and also if cart has items in it they have a button to proceed to checkout.
- added a cart icon into Navbar that shows how many items we have on the cart and also by clicking on it, the cart Item will show up with necessary information and needed actions.
- cart context added with necessary actions.
- the logic is in place that if the user is trying to add a menu from a different chef, they will be warned that they are trying to change chef and if it is what they want they should clear the cart first.
- I also refactored the recipe context to be integrated to our global context, so some of the changes in files are related to that.
- If you want to test navigating to other chefs profiles and test the cart logic, you will need to go to /chefprofile/:userId . The redirecting will happen in brows component that currently I don't have access to. but it will be a simple Link element with correct 'to' attribute. But for now if you want to test it you can use one of the ids below or you can copy a chef's userId from database:

Mina khamesi:605230d1b419c6668cc144e2
Thomas:60522a73b419c6668cc144db
Mina KH :6051fc43bb3a7b17e7408d56

Screen Shot 2021-03-25 at 8 13 58 AM
Screen Shot 2021-03-25 at 8 15 00 AM
Screen Shot 2021-03-25 at 8 15 12 AM

@jessdesr @dproc96

### CODE REVIEW

File: client/src/reducers/CartReducer.js [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+  ADD_TO_CART,
+  DECREASE_COUNT,
+  DELETE_FROM_CART,
+  INCREASE_COUNT,
+  RESET_CART,
+  GET_CHOSEN_CHEF_PROFILE,
+} from "../constants/userConstants";
+
+export const cartInitialState = {
+  cart: localStorage.getItem("cart")
```
Comment (Vaibhav-Kambli) [inline client/src/reducers/CartReducer.js]: This is nice, I will be adding ShippingAddress from localstorage inside the cartInitialState. I will keep the format same inside my cart reducer

File: client/src/components/Dialogs/CartModal.js [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+              variant="contained"
+              className={classes.clearBtn}
+              onClick={() => clearCart(dispatch)}
+            >
+              clear cart
+            </Button>
+          </ListItem>
+          <ListItem alignItems="center">
+            <ListItemText>
+              <Typography variant="h5" align="center">{`chef: ${
```
Comment (dproc96) [inline client/src/components/Dialogs/CartModal.js]: I don't think you need to include "chef:" here

File: client/src/components/Dialogs/CartModal.js [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+              className={classes.img}
+            />
+          </ListItem>
+          <ListItem divider>
+            <ListItemText>
+              <Typography variant="h6">Bio</Typography>
+              <span>{chosenChefProfile && chosenChefProfile.bio}</span>
+            </ListItemText>
+            <ListItemText>
+              <Typography variant="h6">Experts In:</Typography>
```
Comment (dproc96) [inline client/src/components/Dialogs/CartModal.js]: Specialties may be better word choice here

File: client/src/components/Dialogs/CartModal.js [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+
+                    <TableCell align="right"></TableCell>
+                  </TableRow>
+                </TableHead>
+                <TableBody>
+                  {cart.map(({ qty, recipe }) => (
+                    <TableRow key={recipe._id}>
+                      <TableCell component="th" scope="item">
+                        <Avatar
+                          alt="Remy Sharp"
```
Comment (dproc96) [inline client/src/components/Dialogs/CartModal.js]: Should this alt text be the recipe name?

File: client/src/components/Dialogs/CartModal.js [hunk 5]
```diff
@@ -0,0 +1,264 @@
...
+                            </Button>
+                          </ButtonGroup>
+                        </TableCell>
+                      </Hidden>
+                      <Hidden smDown>
+                        <TableCell align="right">{`${
+                          recipe.price
+                        }x${qty}=${addDecimals(
+                          qty * recipe.price
+                        )}`}</TableCell>
```
Comment (dproc96) [inline client/src/components/Dialogs/CartModal.js]: I think you could increase the spacing of this both in the UI and in the code for readability of both

File: client/src/components/Dialogs/CartModal.js [hunk 6]
```diff
@@ -0,0 +1,264 @@
...
+                            <Button
+                              onClick={() =>
+                                qty > 1
+                                  ? decreaseCount(dispatch, {
+                                      id: recipe._id,
+                                      qty: qty - 1,
+                                      chef: recipe.user,
+                                    })
+                                  : removeFromCart(dispatch, recipe._id)
+                              }
```
Comment (dproc96) [inline client/src/components/Dialogs/CartModal.js]: Would be good to refactor these out to handler functions and pass those to onClick, would improve readability

File: client/src/components/Dialogs/CartModal.js [hunk 7]
```diff
@@ -0,0 +1,264 @@
...
+            <ListItemText>
+              <Typography gutterBottom variant="h6">
+                {" "}
+                {addDecimals(
+                  cart.reduce(
+                    (acc, cur) => acc + cur.recipe.price * cur.qty,
+                    0
+                  ),
+                  2
+                )}
```
Comment (dproc96) [inline client/src/components/Dialogs/CartModal.js:249]: This is a nitpick, but it would be good to add a $ here

File: client/src/components/ChefProfile/ChefRecipes.js [hunk 8]
```diff
@@ -1,14 +1,27 @@
...
+  setChefConflictError,
+} from "../../actions/cartActions";
+import { UserContext, UserDispatchContext } from "../../context/UserContext";
 
-const ChefProfile = ({ id }) => {
-  const { recipes } = useContext(RecipeContext);
-  const recipe = recipes.filter((res) => res._id === id)[0];
+const ChefProfile = ({ recipe, isOwner }) => {
+  // const { recipes } = useContext(RecipeContext);
+  // const recipe = recipes.filter((res) => res._id === id)[0];
```
Comment (dproc96) [inline client/src/components/ChefProfile/ChefRecipes.js]: Are these not going to be used anymore? If so they can be deleted?

File: client/src/components/CartIcon.js [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+import { getCartInfo } from "../actions/cartActions";
+
+const StyledBadge = withStyles((theme) => ({
+  badge: {
+    right: -3,
+    top: 13,
+    border: `2px solid ${theme.palette.background.paper}`,
+    padding: "0 4px",
+  },
+}))(Badge);
```
Comment (dproc96) [inline client/src/components/CartIcon.js]: It would be better to be consistent and style this the same way as other components in the makeStyles function

**Review by dproc96 — COMMENTED** [review]
Had a couple of comments, I also wonder if the approach of prototyping so much of this (using test components etc.) is creating more complexity and work for you later than it's saving now. Maybe you can discuss this in your standup but it may be simpler to just build the components directly rather than making proof of concepts.

File: client/src/components/Dialogs/CartModal.js (hunk 7 above)
Comment (MinaKhamesi) [reply client/src/components/Dialogs/CartModal.js:249]: Hey @dproc96 , I have added a $ in one line before this, that's why this comment is not outdated I guess.

Also, I wanted to say that the Test components were not for testing actually, the thing is I was working on recipe and chef profile for this ticket and also for my previous ticket but the actual components were  not written yet, so I had to write a simple Temporary component to do my tickets.  Last night, I finally got the actual components and integrated my code to them. We will have another refactoring in recent future, I thought since I was working with the Test components for the most part, I keep them for now, in case I forgot some edge cases that I took care of before. But after that refactoring, I will get rid of the TestComponents or rather TemporaryComponents completely.

File: client/src/App.js [hunk 10]
```diff
@@ -31,7 +31,8 @@ const Private = () => {
         <Route path="/payment" render={(props) => <Payment {...props} />} />
         <Route path="/test" component={TestComponent} />
         <Route path="/profile" component={CustomerProfile} />
-        <Route path="/chefprofile" component={ChefProfile} />
+        <Route path="/chefprofile/:userId?" component={ChefProfile} />
+        <Route path="/chef/:userId?/recipeId?" component={TestChefProfile} />
```
Comment (dproc96) [inline client/src/App.js:35]:
```suggestion
        <Route path="/chef/:userId?/:recipeId?" component={TestChefProfile} />
```
I had to make this change to get the Chef profile to show, not sure why it worked before the recent changes

**Review by dproc96 — APPROVED** [review]
This looks good, nice work on this. Had one small thing I noticed before you merge

File: client/src/App.js (hunk 10 above)
Comment (Vaibhav-Kambli) [reply client/src/App.js:35]: This is happening because we had Login component on "/" route initially and it was redirecting us, since there is no component rending on "/", it's failing to redirect to proper component. I will have a discussion with the team and render a possible Home component on "/" route that will redirect to proper component based on the type of user logged in

**Comment by jessdesr** [comment]
A few thoughts on the styling:
- The Clear cart button can be inline with the title
- It might make more sense to have the chef contained inside a card in order to group the information (and potentially move the label that it's a chef instead of doing  `chef: thomas`
- I think the recipes are looking somewhat cramped at the moment: perhaps they should be styled to be two lines instead of one, and not bet set up as a table.  I'm thinking:
|
|                                        Recipe Name                                              Price Per Recipe
|      Image                                                                                                                                             Delete Button
|                                         Qty          -     +                                           Total                
|


(ignore this I'll try to draw something)
